lightning/src/util/ser_macros.rs

```diff
@@ -881,19 +903,57 @@ macro_rules! _init_and_read_tlv_stream {
 		$(
 			$crate::_init_tlv_field_var!($field, $fieldty);
 		)*
-
 		$crate::decode_tlv_stream!($reader, {
 			$(($type, $field, $fieldty)),*
 		});
 	}
 }
 
+/// Dummy macro that drops the second argument (which is used by
+/// [`lightning_macros::drop_legacy_field_definition`] to match for legacy fields but isn't needed
+/// in the final code we want to generate).
+#[doc(hidden)]
+#[macro_export]
+macro_rules! _ignore_arg {
+	($field: ident, $fieldty: tt) => {
+		$field
+	};
+}
+
+/// Reads a TLV stream with the given fields to build a struct/enum variant of type `$thing`
+#[doc(hidden)]
+#[macro_export]
+macro_rules! _decode_and_build {
+	($stream: ident, $thing: path, {$(($type: expr, $field: ident, $fieldty: tt)),* $(,)*}) => { {
+		$crate::_init_and_read_len_prefixed_tlv_fields!($stream, {
+			$(($type, $field, $fieldty)),*
+		});
+		// rustc is kinda dumb about unused variable warnings when we declare a variable via an
+		// ident in a macro and then use it in an expr also defined in the same macro call. Thus,
+		// it may generate unused variable warnings for variables that are, in fact, very much
+		// used. Instead, we just blanket ignore unused variables here as it may be useful to write
+		// field names without a _ prefix for legacy fields even if we don't care about the read
+		// value.
+		$(
+			#[allow(unused_variables)]
+			let $field = $crate::_init_tlv_based_struct_field!($field, $fieldty);
+		)*
+		::lightning_macros::drop_legacy_field_definition!($thing {
+			$($field: $crate::_ignore_arg!($field, $fieldty)),*
+		})
+	} }
+}
+
 /// Implements [`Readable`]/[`Writeable`] for a struct storing it as a set of TLVs
 /// If `$fieldty` is `required`, then `$field` is a required field that is not an [`Option`] nor a [`Vec`].
 /// If `$fieldty` is `(default_value, $default)`, then `$field` will be set to `$default` if not present.
 /// If `$fieldty` is `option`, then `$field` is optional field.
 /// If `$fieldty` is `optional_vec`, then `$field` is a [`Vec`], which needs to have its individual elements serialized.
 ///    Note that for `optional_vec` no bytes are written if the vec is empty
+/// If `$fieldty` is `(legacy, $ty, $write)` then, when writing, the function $write will be
+///    called with the object being serialized and a returned `Option` and is written as a TLV if
+///    `Some`. When reading, an optional field of type `$ty` is read (which can be used in later
+///    `default_value` or `static_value` fields).
```

Comment:
Doesn't look like `static_value` is document here but is a the top of the file. `legacy` isn't documented there, though. maybe because we don't want to support it publicly? Could you update these accordingly?

It also isn't clear when the `static_value`s are set, which would affect how it should be used if accessing other TLVs (e.g., in `ChannelDetails`) or other static values for that matter. When reading the macros, it's clear this happens after reading all TLVs and in the order given, but this isn't apparent from the docs, IMO.